soufi/finders/yum.py

```diff
@@ -0,0 +1,215 @@
+# Copyright (c) 2021 Cisco Systems, Inc. and its affiliates
+# All rights reserved.
+
+import abc
+import functools
+import urllib
+
+import repomd
+import requests
+
+from soufi import exceptions, finder
+
+TIMEOUT = 30  # seconds
+
+
+class YumFinder(finder.SourceFinder, metaclass=abc.ABCMeta):
+    """An abstract base class for making Yum-based finders.
+
+    Subclasses of this should provide methods for setting up the repository
+    search lists to pass to __init__.
+
+    The lookup is a 2-stage process:
+        1. Attempt to look up the SRPM directly from the sources repository.
+           In these cases, we use the name only and ignore the version,
+           since it is a very common practice for repositories to only
+           publish repodata for the "current" version, while still
+           providing older sources.
+        2. Attempt to look up the name of the source RPM from the binary
+           repository.  This will catch packages where the source and
+           binary package names do not match.  The version is also ignored
+           in this step, for the same reasons.  Then backtrack and attempt
+           to lookup that package in the source repos.
+    """
+
+    def __init__(self, *args, source_repos=None, binary_repos=None, **kwargs):
+        self.source_repos = source_repos
+        if not self.source_repos:
+            self.source_repos = self._get_source_repos()
+
+        self.binary_repos = binary_repos
+        if not self.binary_repos:
+            self.binary_repos = self._get_binary_repos()
+
+        super().__init__(*args, **kwargs)
+
+    @abc.abstractmethod
+    def _get_source_repos(self):
+        raise NotImplementedError  # pragma: nocover
+
+    @abc.abstractmethod
+    def _get_binary_repos(self):
```

Comment:
These should be public (ie no _) by definition - they're abstract.

Reply:
I went back and forth on this one for a bit:

Are they public, since they NEED to be implemented by subclasses of `YumFinder`?

Or are they private, since they are implementation details of the main `SourceFinder` class, which only exposes one public method, `find()`, and these methods should never be called directly?  

In the end, I let the fact that you created an abstract private method called `SourceFinder._find()` decide the issue for me.  So let me drink in the sweet, sweet irony of you pushing back on this.  Ahhhh...  😆

Reply:
I figured you'd mention that :smile: I screwed that one up, to be fair, and it should be renamed to something else entirely. I used `_find` because `find` is already used.

> Are they public, since they NEED to be implemented by subclasses

Yes, that is the definition of public. This is not C++! We don't do "protected".